### ISSUE TITLE
Improve the word facets to use an informed tokenizer instead of splitting on spaces

### ISSUE THREAD

**Issue by wetneb** [OpenRefine/OpenRefine#2813]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**

The "word facet" that is included in the UI (as a customized text facet) is pretty bare-bones since it just splits strings by spaces: `value.split(" ")`. This means that strings like `"no free view? no review!"` will be split as `["no", "free", "view?", "no", "review!"]`. In this example the trailing punctuation will pollute the facet results.

**Describe the solution you'd like**

GREL could include better tokenizers which would split the text according to established heuristics. There many off-the-shelf tokenizer implementations in Java (for instance in the Lucene project), some of which are language-specific, and others are meant to be sensible catch-alls.

This function would be useful beyond the word facet, as it could obviously be called by users in many other contexts.

**Describe alternatives you've considered**

These heuristics could be installed as Python libraries and imported as a Jython expression - but the workflows to install Python libraries in OpenRefine is cumbersome.

If the dependencies for this are too large, perhaps this is a better job for an extension (especially if many languages are to be included).

**Additional context**
Brought up by @allanaaa in this thread:
https://groups.google.com/forum/#!topic/openrefine-dev/o1Wygcv5Y2Y

**Comment by tfmorris** [comment]
Tokenizers for search and NLP tend to produce results very different from what a data wrangler would want if they were trying to preserve the original form. The current state of the art tokenizers also tend to be NN based, which may be more heavyweight than we need.

**Comment by wetneb** [comment]
> Tokenizers for search and NLP tend to produce results very different from what a data wrangler would want if they were trying to preserve the original form.

Can you expand a bit on this, perhaps with an example? Intuitively a NLP tokenizer would be quite useful in OR, but to be honest I have never needed to split a string into words in OR myself.

**Comment by tfmorris** [comment]
Here's an example from the Stanford NLP page:

> $ cat >sample.txt
> "Oh, no," she's saying, "our $400 blender can't handle something this hard!"
> $ java edu.stanford.nlp.process.PTBTokenizer sample.txt
> ``
> Oh
> ,
> no
> ,
> ''
> she
> 's
> saying
> ,
> ``
> our
> $
> 400
> blender
> ca
> n't
> handle
> something
> this
> hard
> !
> ''
> PTBTokenizer tokenized 23 tokens at 370.97 tokens per second.

Not all of them are that bad and there is a huge variety, but just wanted to flag that the needs of various NLP algorithms aren't necessarily aligned with the needs of OpenRefine users. They're also all, except for the very simple ones, language specific.

**Comment by tfmorris** [comment]
This is largely a duplicate, but superset, of #557.

**Comment by tfmorris** [comment]
Folks seem to have settled on UAX29 as the preferred tokenization algorithm for both NLP & search, which makes sense since the Unicode Consortium should know what they're doing in this space.

It's implemented by this Lucene tokenizer https://lucene.apache.org/core/3_5_0/api/core/org/apache/lucene/analysis/standard/StandardTokenizerImpl.html

**Comment by thadguidry** [comment]
@tfmorris Agree Tom, and UAX29 can help considerably, but depending on the work a user is doing (If working with Latin based languages or not) then I think additionally, it would be good for OpenRefine users to have control over some of the boundary rules within it, for instance `Grapheme_Extend = Yes` and others like:
> An important rule for the default word and sentence specifications ignores Extend and Format characters. The main purpose of this rule is to always treat a grapheme cluster as a single character—that is, to not break a single grapheme cluster across two higher-level segments.

 Basically, the boundary rules will differ a bit depending on OpenRefine users particular needs...so let's continue to give them the interface controls they might potentially need.  When the time comes perhaps @lozanaross can help us with an overlay prototype for facet controls, which might be extension points or not.

**Comment by tfmorris** [comment]
#5551 which closed #557 has significantly improved the situation. The word separator is now the Unicode pattern \W (equivalent to `[^\w]` where `\w` is defined as `[\p{Alpha}\p{gc=Mn}\p{gc=Me}\p{gc=Mc}\p{Digit}\p{gc=Pc}\p{IsJoin_Control}]`
https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html

**Comment by thadguidry** [comment]
Yeah, the "word facet" is much better and standardized around `[^\w]` which makes sense to me.
I'd be in favor of closing this issue.
For folks that need more control on word splitting, then they can have their day with a custom GREL text facet and use the power of RegEx within our `value.split()`.

@wetneb ?  Shall you close this?

**Comment by wetneb** [comment]
I think our custom clustering feature would be all the more useful if we had some more advanced tokenizers available as GREL functions (which could also be used for a word facet), so to me the need is still valid.
